File: nse_catalog.py
"""
nse_catalog.py - Invest IQ Official NSE Equity Catalog & Search Engine

Indexes all 2,200+ active equity stocks listed on the National Stock Exchange of India (NSE).
Features:
- Bundled offline catalog for instantaneous sub-millisecond queries.
- Multi-field search (ticker symbol, company name, ISIN).
- Industry/sector classification (IT, Banking, Pharma, Auto, Energy, Metals, FMCG, Defense, etc.).
- Fast server-side pagination with complete page metadata.
"""

import os
import json
import sqlite3
import logging
from typing import Dict, List, Any, Optional, Tuple

logger = logging.getLogger("InvestIQ.NSECatalog")

# Base directory
BASE_DIR = os.path.dirname(__file__)
CACHE_DIR = os.path.join(BASE_DIR, "data_cache")
RAW_CACHE_FILE = os.path.join(CACHE_DIR, "nse_equities_raw.json")
CATALOG_DB_PATH = os.path.join(BASE_DIR, "investiq.db")

# Keyword-based sector categorization mapping for NSE companies
SECTOR_KEYWORDS = [
    ("Information Technology", ["technolog", "software", "infotech", "systems", "tech", "digital", "data", "it solutions", "mindtree", "wipro", "infosys", "tcs", "cyient", "kpit", "persistent"]),
    ("Banking & Financials", ["bank", "finance", "capital", "housing finance", "holding", "invest", "securities", "finserv", "bajaj fin", "credit", "asset", "insurance"]),
    ("Healthcare & Pharma", ["pharma", "laborator", "health", "hospital", "lifescience", "drug", "medic", "biotech", "chem", "sun pharma", "cipla", "dr. reddy", "lupin", "aurobindo", "divi"]),
    ("Automobile & Auto Parts", ["motor", "auto", "tyre", "wheel", "battery", "gear", "clutch", "chassis", "forging", "maruti", "mahindra", "tata motor", "eicher", "tvs", "bajaj auto", "hero"]),
    ("Energy, Oil & Power", ["power", "energy", "petro", "oil", "gas", "refin", "thermal", "solar", "renewable", "electric", "ongc", "ntpc", "iocl", "bpcl", "coal", "adani green", "tata power"]),
    ("Metals & Mining", ["steel", "iron", "metal", "mining", "aluminum", "zinc", "copper", "alloy", "jsw", "jindal", "tata steel", "vedanta", "hindalco", "sail", "nmdc"]),
    ("FMCG & Consumer Goods", ["food", "beverag", "consumer", "agro", "tea", "sugar", "brewer", "dairy", "oil", "flour", "retail", "hindustan unilever", "itc", "nestle", "britannia", "marico", "dabur"]),
    ("Infrastructure & Construction", ["infra", "construct", "cement", "engineer", "build", "project", "pipe", "realty", "estate", "larsen", "ultratech", "dlf", "godrej prop"]),
    ("Telecommunications", ["airtel", "telecom", "tele-", "network", "optic", "broadband", "tata comm", "vodafone"]),
    ("Defense & Aerospace", ["defense", "defence", "aerospace", "aviation", "hal", "bharat electron", "mazagon", "cochin shipyard", "paras", "zen tech"]),
    ("Chemicals & Fertilizers", ["chem", "fertiliz", "carbon", "pigment", "organic", "polymer", "pesticide", "aarti", "deepak", "tata chem", "upl"]),
    ("Media & Entertainment", ["media", "network", "film", "cinema", "entertain", "broadcast", "multiplex", "pvr", "inox", "zeel", "sun tv"]),
]

# Market cap hints based on major benchmarks
LARGE_CAP_HINTS = {
    "RELIANCE", "TCS", "HDFCBANK", "INFY", "ICICIBANK", "BHARTIARTL", "SBIN", "ITC",
    "HINDUNILVR", "LT", "KOTAKBANK", "AXISBANK", "BAJFINANCE", "BAJAJFINSV", "SUNPHARMA",
    "TITAN", "MARUTI", "TATAMOTORS", "TATASTEEL", "NTPC", "ONGC", "POWERGRID", "COALINDIA",
    "ULTRACEMCO", "M&M", "ADANIENT", "ADANIPORTS", "HCLTECH", "WIPRO", "TECHM", "ASIANPAINT",
    "BAJAJ-AUTO", "HEROMOTOCO", "EICHERMOT", "NESTLEIND", "BRITANNIA", "DIVISLAB", "CIPLA",
    "DRREDDY", "APOLLOHOSP", "INDUSINDBK", "JSWSTEEL", "HINDALCO", "GRASIM", "SHREECEM",
    "BPCL", "IOC", "TATAPOWER", "BEL", "HAL", "VEDL", "ZOMATO", "PAYTM", "JIOFIN"
}


def _infer_sector(symbol: str, company_name: str) -> str:
    """Infer sector classification from company name and symbol."""
    text = f"{symbol} {company_name}".lower()
    for sector_name, keywords in SECTOR_KEYWORDS:
        for kw in keywords:
            if kw in text:
                return sector_name
    return "Diversified & General"


def _infer_cap(symbol: str) -> str:
    """Infer market cap category."""
    clean_sym = symbol.replace(".NS", "").upper()
    if clean_sym in LARGE_CAP_HINTS:
        return "Large Cap"
    return "Mid & Small Cap"


class NSECatalog:
    """In-memory and SQLite-backed catalog of all listed NSE equity stocks."""

    def __init__(self, db_path: str = CATALOG_DB_PATH):
        self.db_path = db_path
        self._memory_cache: List[Dict[str, Any]] = []
        self._symbol_map: Dict[str, Dict[str, Any]] = {}
        self.init_catalog()

    def init_catalog(self) -> None:
        """Initialize SQLite catalog table and populate if needed."""
        conn = sqlite3.connect(self.db_path)
        conn.execute("PRAGMA journal_mode=WAL")
        cursor = conn.cursor()
        cursor.execute("""
            CREATE TABLE IF NOT EXISTS nse_catalog (
                symbol TEXT PRIMARY KEY,
                name TEXT NOT NULL,
                series TEXT NOT NULL,
                isin TEXT,
                sector TEXT NOT NULL,
                cap_category TEXT NOT NULL,
                listing_date TEXT
            );
        """)
        cursor.execute("CREATE INDEX IF NOT EXISTS idx_nse_symbol ON nse_catalog(symbol);")
        cursor.execute("CREATE INDEX IF NOT EXISTS idx_nse_sector ON nse_catalog(sector);")
        conn.commit()

        # Check if table already populated
        cursor.execute("SELECT COUNT(*) FROM nse_catalog")
        count = cursor.fetchone()[0]

        if count < 1000:
            logger.info("Populating nse_catalog table from equity data...")
            self._populate_from_raw_cache(conn)

        # Load into memory cache for instant <1ms filtering
        cursor.execute("""
            SELECT symbol, name, series, isin, sector, cap_category, listing_date 
            FROM nse_catalog 
            ORDER BY 
                CASE 
                    WHEN symbol IN ('RELIANCE', 'TCS', 'HDFCBANK', 'INFY', 'ICICIBANK', 'SBIN', 'BHARTIARTL', 'TATAMOTORS', 'ITC', 'KOTAKBANK', 'LT', 'HINDUNILVR') THEN 0 
                    WHEN cap_category = 'Large Cap' THEN 1 
                    ELSE 2 
                END, 
                symbol ASC
        """)
        rows = cursor.fetchall()
        self._memory_cache = []
        self._symbol_map = {}
        for r in rows:
            entry = {
                "symbol": r[0],
                "name": r[1],
                "series": r[2],
                "isin": r[3],
                "sector": r[4],
                "cap": r[5],
                "listing_date": r[6],
            }
            self._memory_cache.append(entry)
            self._symbol_map[r[0]] = entry
            # Also map with .NS suffix for seamless lookup
            self._symbol_map[f"{r[0]}.NS"] = entry

        conn.close()
        logger.info("NSECatalog initialized with %d active NSE stocks.", len(self._memory_cache))

    def _populate_from_raw_cache(self, conn: sqlite3.Connection) -> None:
        """Populate database from raw cached JSON or downloaded CSV."""
        records = []
        if os.path.exists(RAW_CACHE_FILE):
            try:
                with open(RAW_CACHE_FILE, "r", encoding="utf-8") as f:
                    records = json.load(f)
            except Exception as e:
                logger.error("Failed to read %s: %s", RAW_CACHE_FILE, e)

        if not records:
            # Fallback to downloading directly
            try:
                import urllib.request
                import csv
                url = "https://archives.nseindia.com/content/equities/EQUITY_L.csv"
                req = urllib.request.Request(url, headers={"User-Agent": "Mozilla/5.0"})
                with urllib.request.urlopen(req, timeout=15) as res:
                    lines = res.read().decode("utf-8").splitlines()
                    reader = csv.DictReader(lines)
                    for r in reader:
                        clean_r = {k.strip(): v.strip() for k, v in r.items() if k}
                        if clean_r.get("SERIES") == "EQ":
                            records.append(clean_r)
            except Exception as e:
                logger.warning("Could not download live NSE equity master: %s. Using default list.", e)

        if records:
            cursor = conn.cursor()
            for r in records:
                sym = r.get("SYMBOL", "").strip().upper()
                name = r.get("NAME OF COMPANY", sym).strip()
                series = r.get("SERIES", "EQ").strip()
                isin = r.get("ISIN NUMBER", "").strip()
                list_date = r.get("DATE OF LISTING", "").strip()
                sector = _infer_sector(sym, name)
                cap = _infer_cap(sym)

                cursor.execute("""
                    INSERT OR REPLACE INTO nse_catalog
                    (symbol, name, series, isin, sector, cap_category, listing_date)
                    VALUES (?, ?, ?, ?, ?, ?, ?)
                """, (sym, name, series, isin, sector, cap, list_date))
            conn.commit()

    def get_stock(self, symbol: str) -> Optional[Dict[str, Any]]:
        """Instant lookup for a stock by symbol (e.g. 'RELIANCE' or 'RELIANCE.NS')."""
        clean = symbol.strip().upper()
        return self._symbol_map.get(clean) or self._symbol_map.get(clean.replace(".NS", ""))

    def get_sectors(self) -> List[str]:
        """Return list of distinct available sectors."""
        sectors = sorted(list({s["sector"] for s in self._memory_cache}))
        return sectors

    def search_stocks(
        self,
        query: str = "",
        sector: Optional[str] = None,
        cap: Optional[str] = None,
        page: int = 1,
        per_page: int = 25,
    ) -> Dict[str, Any]:
        """
        Fast server-side paginated search across all 2,200+ NSE stocks.
        Returns matching stocks, total count, and page navigation metadata.
        """
        q = query.strip().upper()
        filtered = self._memory_cache

        # Filter by text search
        if q:
            filtered = [
                s for s in filtered
                if q in s["symbol"] or q in s["name"].upper() or (s["isin"] and q in s["isin"])
            ]

        # Filter by sector
        if sector and sector != "ALL":
            filtered = [s for s in filtered if s["sector"].lower() == sector.lower()]

        # Filter by cap
        if cap and cap != "ALL":
            filtered = [s for s in filtered if s["cap"].lower() == cap.lower()]

        total_records = len(filtered)
        per_page = max(5, min(100, per_page))
        total_pages = max(1, (total_records + per_page - 1) // per_page)
        current_page = max(1, min(page, total_pages))

        start_idx = (current_page - 1) * per_page
        end_idx = start_idx + per_page
        page_items = filtered[start_idx:end_idx]

        return {
            "stocks": page_items,
            "total": total_records,
            "total_count": total_records,
            "page": current_page,
            "per_page": per_page,
            "total_pages": total_pages,
            "has_prev": current_page > 1,
            "has_next": current_page < total_pages,
            "prev_page": current_page - 1,
            "next_page": current_page + 1,
        }


# Global singleton instance
default_nse_catalog = NSECatalog()

File: frontend/src/lib/nse-catalog.ts
import nseRawCatalog from "../../data/nse_catalog.json";

export interface NseStock {
  symbol: string;
  nseSymbol: string; // Ticker formatted for upstream quote engine, e.g. RELIANCE.NS
  name: string;
  series: string;
  isin: string;
  sector: string;
  category: "large-cap" | "mid-cap" | "small-cap" | "index";
}

// Known Sector overrides for Premier Indian Stocks
const SECTOR_OVERRIDES: Record<string, { sector: string; category: "large-cap" | "mid-cap" | "small-cap" }> = {
  RELIANCE: { sector: "Energy & Conglomerate", category: "large-cap" },
  TCS: { sector: "Information Technology", category: "large-cap" },
  HDFCBANK: { sector: "Banking & Financials", category: "large-cap" },
  INFY: { sector: "Information Technology", category: "large-cap" },
  ICICIBANK: { sector: "Banking & Financials", category: "large-cap" },
  BHARTIARTL: { sector: "Telecommunications", category: "large-cap" },
  SBIN: { sector: "PSU Banking", category: "large-cap" },
  ITC: { sector: "FMCG & Consumer", category: "large-cap" },
  HINDUNILVR: { sector: "FMCG & Consumer", category: "large-cap" },
  LT: { sector: "Infrastructure & Engineering", category: "large-cap" },
  KOTAKBANK: { sector: "Banking & Financials", category: "large-cap" },
  AXISBANK: { sector: "Banking & Financials", category: "large-cap" },
  BAJFINANCE: { sector: "NBFC & Financial Services", category: "large-cap" },
  BAJAJFINSV: { sector: "Financial Services", category: "large-cap" },
  SUNPHARMA: { sector: "Healthcare & Pharma", category: "large-cap" },
  TITAN: { sector: "Consumer & Jewelry", category: "large-cap" },
  MARUTI: { sector: "Automobile & EV", category: "large-cap" },
  TMPV: { sector: "Automobile & EV", category: "large-cap" },
  TMCV: { sector: "Automobile & Commercial", category: "large-cap" },
  TATAMOTORS: { sector: "Automobile & EV", category: "large-cap" },
  TATASTEEL: { sector: "Metals & Mining", category: "large-cap" },
  NTPC: { sector: "Power & Energy (PSU)", category: "large-cap" },
  ONGC: { sector: "Oil & Gas (PSU)", category: "large-cap" },
  POWERGRID: { sector: "Power Transmission (PSU)", category: "large-cap" },
  COALINDIA: { sector: "Mining & Energy (PSU)", category: "large-cap" },
  ULTRACEMCO: { sector: "Building Materials & Cement", category: "large-cap" },
  "M&M": { sector: "Automobile & Farm Equipment", category: "large-cap" },
  ADANIENT: { sector: "Conglomerate & Infrastructure", category: "large-cap" },
  ADANIPORTS: { sector: "Ports & Logistics", category: "large-cap" },
  HCLTECH: { sector: "Information Technology", category: "large-cap" },
  WIPRO: { sector: "Information Technology", category: "large-cap" },
  ASIANPAINT: { sector: "Paints & Home Decor", category: "large-cap" },
  TECHM: { sector: "Information Technology", category: "large-cap" },
  GRASIM: { sector: "Chemicals & Textiles", category: "large-cap" },
  HINDALCO: { sector: "Aluminium & Metals", category: "large-cap" },
  CIPLA: { sector: "Healthcare & Pharma", category: "large-cap" },
  DRREDDY: { sector: "Healthcare & Pharma", category: "large-cap" },
  JSWSTEEL: { sector: "Metals & Mining", category: "large-cap" },
  HEROMOTOCO: { sector: "Automobile & Two-Wheelers", category: "large-cap" },
  "BAJAJ-AUTO": { sector: "Automobile & Two-Wheelers", category: "large-cap" },
  EICHERMOT: { sector: "Automobile (Royal Enfield)", category: "large-cap" },
  DIVISLAB: { sector: "Healthcare & Pharma", category: "large-cap" },
  BRITANNIA: { sector: "FMCG & Foods", category: "large-cap" },
  NESTLEIND: { sector: "FMCG & Consumer", category: "large-cap" },
  INDUSINDBK: { sector: "Banking & Financials", category: "large-cap" },
  TATACONSUM: { sector: "FMCG & Consumer", category: "large-cap" },
  APOLLOHOSP: { sector: "Healthcare & Hospitals", category: "large-cap" },
  SHRIRAMFIN: { sector: "NBFC & Lending", category: "large-cap" },
  BPCL: { sector: "Oil & Gas (PSU)", category: "large-cap" },
  LTIM: { sector: "Information Technology", category: "large-cap" },
  TRENT: { sector: "Retail & Consumer (Zudio)", category: "large-cap" },
  BEL: { sector: "Defense & Aerospace (PSU)", category: "large-cap" },
  HAL: { sector: "Defense & Aerospace (PSU)", category: "large-cap" },
  MAZDOCK: { sector: "Defense & Shipbuilding (PSU)", category: "mid-cap" },
  COCHINSHIP: { sector: "Defense & Shipbuilding (PSU)", category: "mid-cap" },
  BDL: { sector: "Defense & Missiles (PSU)", category: "mid-cap" },
  IRFC: { sector: "Railways & Infrastructure (PSU)", category: "large-cap" },
  RVNL: { sector: "Railways & Infrastructure (PSU)", category: "mid-cap" },
  IRCTC: { sector: "Railways & Tourism (PSU)", category: "mid-cap" },
  RAILTEL: { sector: "Railways & Telecom (PSU)", category: "mid-cap" },
  SUZLON: { sector: "Renewable Energy & Wind", category: "mid-cap" },
  PAYTM: { sector: "Fintech & Payments", category: "mid-cap" },
  YESBANK: { sector: "Banking & Financials", category: "mid-cap" },
  ETERNAL: { sector: "Consumer Internet & Quick Commerce (Zomato/Blinkit)", category: "large-cap" },
  JIOFIN: { sector: "Fintech & Financials", category: "large-cap" },
  BSE: { sector: "Financial Market Infrastructure", category: "mid-cap" },
  CDSL: { sector: "Financial Market Infrastructure", category: "mid-cap" },
  TATAPOWER: { sector: "Energy & Power", category: "large-cap" },
  VEDL: { sector: "Metals & Mining", category: "large-cap" },
  POLYCAB: { sector: "Cables & Electricals", category: "large-cap" },
  DIXON: { sector: "Electronics Manufacturing", category: "mid-cap" },
  KPITTECH: { sector: "Automotive Software & EV Tech", category: "mid-cap" },
  PERSISTENT: { sector: "Information Technology", category: "mid-cap" },
  IDFCFIRSTB: { sector: "Banking & Financials", category: "mid-cap" },
  BANKBARODA: { sector: "PSU Banking", category: "large-cap" },
  PNB: { sector: "PSU Banking", category: "large-cap" },
};

function inferSector(name: string, symbol: string): { sector: string; category: "large-cap" | "mid-cap" | "small-cap" } {
  const sym = symbol.toUpperCase().replace(/\.NS$/, "");
  if (SECTOR_OVERRIDES[sym]) {
    return SECTOR_OVERRIDES[sym];
  }

  const n = name.toLowerCase();

  if (n.includes("bank") || n.includes("finance") || n.includes("capital") || n.includes("credit") || n.includes("housing") || n.includes("leasing") || n.includes("securities")) {
    return { sector: "Banking & Financials", category: "mid-cap" };
  }
  if (n.includes("tech") || n.includes("software") || n.includes("infotech") || n.includes("digital") || n.includes("system") || n.includes("data") || n.includes("cyber")) {
    return { sector: "Information Technology", category: "mid-cap" };
  }
  if (n.includes("pharma") || n.includes("health") || n.includes("drug") || n.includes("lab") || n.includes("med") || n.includes("bio") || n.includes("hospital") || n.includes("care")) {
    return { sector: "Healthcare & Pharma", category: "mid-cap" };
  }
  if (n.includes("motor") || n.includes("auto") || n.includes("tyre") || n.includes("wheel") || n.includes("vehicle") || n.includes("engine") || n.includes("clutch")) {
    return { sector: "Automobile & EV", category: "mid-cap" };
  }
  if (n.includes("defense") || n.includes("defence") || n.includes("aero") || n.includes("ship") || n.includes("dynamic") || n.includes("armament")) {
    return { sector: "Defense & Aerospace", category: "mid-cap" };
  }
  if (n.includes("rail") || n.includes("infra") || n.includes("engineering") || n.includes("construct") || n.includes("build") || n.includes("cement") || n.includes("project")) {
    return { sector: "Railways & Infrastructure", category: "mid-cap" };
  }
  if (n.includes("power") || n.includes("energy") || n.includes("solar") || n.includes("wind") || n.includes("petrol") || n.includes("gas") || n.includes("oil") || n.includes("electric")) {
    return { sector: "Energy & Power", category: "mid-cap" };
  }
  if (n.includes("steel") || n.includes("metal") || n.includes("iron") || n.includes("zinc") || n.includes("copper") || n.includes("aluminium") || n.includes("alloy") || n.includes("mining")) {
    return { sector: "Metals & Mining", category: "mid-cap" };
  }
  if (n.includes("food") || n.includes("sugar") || n.includes("tea") || n.includes("coffee") || n.includes("dairy") || n.includes("beverag") || n.includes("agro") || n.includes("consumer") || n.includes("fmcg")) {
    return { sector: "FMCG & Consumer", category: "mid-cap" };
  }
  if (n.includes("chem") || n.includes("polymer") || n.includes("fertil") || n.includes("plastic") || n.includes("pigment") || n.includes("color")) {
    return { sector: "Chemicals & Materials", category: "mid-cap" };
  }
  if (n.includes("textile") || n.includes("cotton") || n.includes("silk") || n.includes("yarn") || n.includes("spin") || n.includes("garment") || n.includes("fashion") || n.includes("apparel")) {
    return { sector: "Textiles & Apparel", category: "small-cap" };
  }
  if (n.includes("telecom") || n.includes("cable") || n.includes("network") || n.includes("communication")) {
    return { sector: "Telecommunications", category: "mid-cap" };
  }

  return { sector: "Diversified Equities", category: "small-cap" };
}

// Build processed indexed catalog
interface RawEntry {
  SYMBOL: string;
  "NAME OF COMPANY": string;
  SERIES: string;
  "ISIN NUMBER"?: string;
}

const nseStocksList: NseStock[] = (nseRawCatalog as RawEntry[]).map((row) => {
  const symbol = row.SYMBOL.trim();
  const name = row["NAME OF COMPANY"]?.trim() || symbol;
  const inferred = inferSector(name, symbol);
  return {
    symbol,
    nseSymbol: `${symbol}.NS`,
    name,
    series: row.SERIES?.trim() || "EQ",
    isin: row["ISIN NUMBER"]?.trim() || "",
    sector: inferred.sector,
    category: inferred.category,
  };
});

// Fast lookup map
const nseStocksBySymbol = new Map<string, NseStock>();
for (const stock of nseStocksList) {
  nseStocksBySymbol.set(stock.symbol.toUpperCase(), stock);
  nseStocksBySymbol.set(stock.nseSymbol.toUpperCase(), stock);
}

// Aliases for seamless search
const ALIASES: Record<string, string> = {
  ZOMATO: "ETERNAL",
  BLINKIT: "ETERNAL",
  TATAMOTORS: "TMPV",
  "TATA MOTORS": "TMPV",
};

export function getAllNseStocks(): NseStock[] {
  return nseStocksList;
}

export function getNseStock(symbolOrNse: string): NseStock | undefined {
  const clean = symbolOrNse.trim().toUpperCase();
  const direct = nseStocksBySymbol.get(clean);
  if (direct) return direct;

  const aliased = ALIASES[clean];
  if (aliased) {
    return nseStocksBySymbol.get(aliased) || nseStocksBySymbol.get(`${aliased}.NS`);
  }

  const withoutNs = clean.replace(/\.NS$/, "");
  return nseStocksBySymbol.get(withoutNs);
}

export function searchNseStocks(
  query: string,
  options?: { sector?: string; category?: string; limit?: number }
): NseStock[] {
  const q = query.trim().toUpperCase();
  const limit = options?.limit || 50;
  const sectorFilter = options?.sector;
  const categoryFilter = options?.category;
  const aliasMatch = ALIASES[q] || (q.includes("ZOMATO") ? "ETERNAL" : q.includes("TATA MOTOR") ? "TMPV" : null);

  if (!q && !sectorFilter && !categoryFilter) {
    return nseStocksList.slice(0, limit);
  }

  const results: NseStock[] = [];

  for (const stock of nseStocksList) {
    if (sectorFilter && stock.sector !== sectorFilter) continue;
    if (categoryFilter && stock.category !== categoryFilter) continue;

    if (!q) {
      results.push(stock);
      if (results.length >= limit) break;
      continue;
    }

    const sym = stock.symbol.toUpperCase();
    const name = stock.name.toUpperCase();
    const sector = stock.sector.toUpperCase();

    // Prioritize alias matches and exact/prefix matches
    if (aliasMatch && sym === aliasMatch) {
      results.unshift(stock);
    } else if (sym === q || sym.startsWith(q)) {
      results.unshift(stock);
    } else if (sym.includes(q) || name.includes(q) || sector.includes(q)) {
      results.push(stock);
    }

    if (results.length >= limit * 2) break;
  }

  return results.slice(0, limit);
}

// Benchmark indices
export const NSE_BENCHMARKS = [
  { symbol: "^NSEI", nseSymbol: "^NSEI", name: "NIFTY 50", sector: "Indian Benchmark Index", category: "index" as const },
  { symbol: "^BSESN", nseSymbol: "^BSESN", name: "BSE SENSEX", sector: "Indian Benchmark Index", category: "index" as const },
  { symbol: "^NSEBANK", nseSymbol: "^NSEBANK", name: "BANK NIFTY", sector: "Banking Index", category: "index" as const },
  { symbol: "^CNXIT", nseSymbol: "^CNXIT", name: "NIFTY IT", sector: "IT Index", category: "index" as const },
];

// Top 50 Premier Indian Equities for instant loading & live feeds
export const TOP_INDIAN_STOCKS = [
  "RELIANCE.NS",
  "TCS.NS",
  "HDFCBANK.NS",
  "INFY.NS",
  "ICICIBANK.NS",
  "BHARTIARTL.NS",
  "SBIN.NS",
  "ITC.NS",
  "HINDUNILVR.NS",
  "LT.NS",
  "KOTAKBANK.NS",
  "AXISBANK.NS",
  "BAJFINANCE.NS",
  "MARUTI.NS",
  "SUNPHARMA.NS",
  "TITAN.NS",
  "TMPV.NS",
  "TATASTEEL.NS",
  "NTPC.NS",
  "ONGC.NS",
  "POWERGRID.NS",
  "COALINDIA.NS",
  "ULTRACEMCO.NS",
  "M&M.NS",
  "ADANIENT.NS",
  "ADANIPORTS.NS",
  "HCLTECH.NS",
  "WIPRO.NS",
  "ASIANPAINT.NS",
  "HAL.NS",
  "BEL.NS",
  "IRFC.NS",
  "RVNL.NS",
  "SUZLON.NS",
  "ETERNAL.NS", // Zomato/Blinkit parent
  "JIOFIN.NS",
  "BSE.NS",
  "CDSL.NS",
  "TATAPOWER.NS",
  "VEDL.NS",
  "COCHINSHIP.NS",
  "MAZDOCK.NS",
  "DIXON.NS",
  "POLYCAB.NS",
  "PERSISTENT.NS",
  "KPITTECH.NS",
  "IDFCFIRSTB.NS",
  "BANKBARODA.NS",
  "PAYTM.NS",
  "YESBANK.NS",
];

// Import Global Catalog for Multi-Market Support
import {
  TOP_GLOBAL_STOCKS,
  GLOBAL_INDICES,
  GLOBAL_STOCKS_MAP,
  isGlobalSymbol,
  GlobalStock,
} from "./global-catalog";

export { TOP_GLOBAL_STOCKS, GLOBAL_INDICES, GLOBAL_STOCKS_MAP, isGlobalSymbol };
export type { GlobalStock };

export interface UnifiedStock {
  symbol: string;
  quoteSymbol: string;
  name: string;
  market: "NSE" | "GLOBAL" | "INDEX";
  currency: "INR" | "USD";
  series?: string;
  isin?: string;
  sector: string;
  category: "large-cap" | "mid-cap" | "small-cap" | "tech" | "international" | "index";
  exchange?: string;
}

// Build unified list of 2,400+ stocks (2,298 NSE + 100+ Global & Indices)
const unifiedStocksList: UnifiedStock[] = [
  ...GLOBAL_INDICES.map((idx): UnifiedStock => ({
    symbol: idx.symbol,
    quoteSymbol: idx.quoteSymbol,
    name: idx.name,
    market: "INDEX",
    currency: "USD",
    sector: idx.sector,
    category: idx.category,
    exchange: idx.exchange,
  })),
  ...TOP_GLOBAL_STOCKS.map((g): UnifiedStock => ({
    symbol: g.symbol,
    quoteSymbol: g.quoteSymbol,
    name: g.name,
    market: "GLOBAL",
    currency: "USD",
    sector: g.sector,
    category: g.category,
    exchange: g.exchange,
  })),
  ...nseStocksList.map((n): UnifiedStock => ({
    symbol: n.symbol,
    quoteSymbol: n.nseSymbol,
    name: n.name,
    market: "NSE",
    currency: "INR",
    series: n.series,
    isin: n.isin,
    sector: n.sector,
    category: n.category,
    exchange: "NSE",
  })),
];

const unifiedMap = new Map<string, UnifiedStock>();
for (const s of unifiedStocksList) {
  unifiedMap.set(s.symbol.toUpperCase(), s);
  unifiedMap.set(s.quoteSymbol.toUpperCase(), s);
}

export function getAllUnifiedStocks(): UnifiedStock[] {
  return unifiedStocksList;
}

export function getUnifiedStock(sym: string): UnifiedStock | undefined {
  const clean = sym.trim().toUpperCase();
  const direct = unifiedMap.get(clean);
  if (direct) return direct;

  const withoutNs = clean.replace(/\.NS$/, "").replace(/\.BO$/, "");
  const directWithoutNs = unifiedMap.get(withoutNs);
  if (directWithoutNs) return directWithoutNs;

  const aliased = ALIASES[clean] || ALIASES[withoutNs];
  if (aliased) {
    return unifiedMap.get(aliased) || unifiedMap.get(`${aliased}.NS`);
  }

  return undefined;
}

export function searchAllMarkets(
  query: string,
  options?: {
    market?: "all" | "nse" | "global" | "index";
    sector?: string;
    category?: string;
    page?: number;
    limit?: number;
  }
): {
  items: UnifiedStock[];
  total: number;
  page: number;
  totalPages: number;
} {
  const q = (query || "").trim().toUpperCase();
  const marketFilter = options?.market || "all";
  const sectorFilter = options?.sector;
  const categoryFilter = options?.category;
  const page = Math.max(1, options?.page || 1);
  const limit = Math.max(1, Math.min(100, options?.limit || 30));

  let filtered = unifiedStocksList;

  // Filter by market
  if (marketFilter === "nse") {
    filtered = filtered.filter((s) => s.market === "NSE");
  } else if (marketFilter === "global") {
    filtered = filtered.filter((s) => s.market === "GLOBAL" || s.market === "INDEX");
  } else if (marketFilter === "index") {
    filtered = filtered.filter((s) => s.market === "INDEX");
  }

  // Filter by sector
  if (sectorFilter && sectorFilter !== "all") {
    filtered = filtered.filter(
      (s) => s.sector.toLowerCase().includes(sectorFilter.toLowerCase())
    );
  }

  // Filter by category
  if (categoryFilter && categoryFilter !== "all") {
    filtered = filtered.filter((s) => s.category === categoryFilter);
  }

  // Search query
  if (q) {
    const aliasMatch = ALIASES[q] || (q.includes("ZOMATO") ? "ETERNAL" : q.includes("TATA MOTOR") ? "TMPV" : null);

    const exact: UnifiedStock[] = [];
    const prefix: UnifiedStock[] = [];
    const substring: UnifiedStock[] = [];

    for (const s of filtered) {
      const sym = s.symbol.toUpperCase();
      const name = s.name.toUpperCase();
      const sector = s.sector.toUpperCase();

      if (aliasMatch && sym === aliasMatch) {
        exact.push(s);
      } else if (sym === q) {
        exact.push(s);
      } else if (sym.startsWith(q) || name.startsWith(q)) {
        prefix.push(s);
      } else if (sym.includes(q) || name.includes(q) || sector.includes(q)) {
        substring.push(s);
      }
    }

    filtered = [...exact, ...prefix, ...substring];
  }

  const total = filtered.length;
  const totalPages = Math.ceil(total / limit) || 1;
  const startIndex = (page - 1) * limit;
  const items = filtered.slice(startIndex, startIndex + limit);

  return {
    items,
    total,
    page,
    totalPages,
  };
}
